File: src/test/java/org/trp/shincolle/entity/base/SummonBrainDecisionResolverTest.java
package org.trp.shincolle.entity.base;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

class SummonBrainDecisionResolverTest {

    @Test
    void attackShouldRequireAliveTarget() {
        assertTrue(SummonBrainDecisionResolver.shouldAttack(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(true)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldAttack(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(false)
                .build()));
    }

    @Test
    void chaseShouldOnlyContinueWhileTargetIsAliveAndOutOfRange() {
        assertTrue(SummonBrainDecisionResolver.shouldChaseAttackTarget(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(true)
                .targetDistanceSq(16.01D)
                .attackRangeSq(16.0D)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldChaseAttackTarget(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(true)
                .targetDistanceSq(16.0D)
                .attackRangeSq(16.0D)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldChaseAttackTarget(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(false)
                .targetDistanceSq(20.0D)
                .attackRangeSq(16.0D)
                .build()));
    }

    @Test
    void performAttackShouldRequireAliveTargetCooldownAndInRange() {
        assertTrue(SummonBrainDecisionResolver.shouldPerformAttack(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(true)
                .targetDistanceSq(16.0D)
                .attackRangeSq(16.0D)
                .attackDelay(0)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldPerformAttack(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(true)
                .targetDistanceSq(16.01D)
                .attackRangeSq(16.0D)
                .attackDelay(0)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldPerformAttack(stateBuilder()
                .hasAttackTarget(true)
                .hasAliveAttackTarget(true)
                .targetDistanceSq(16.0D)
                .attackRangeSq(16.0D)
                .attackDelay(1)
                .build()));
    }

    @Test
    void followCarrierShouldRequireAliveCarrierDistanceAndNoAttackTarget() {
        assertTrue(SummonBrainDecisionResolver.shouldFollowCarrier(stateBuilder()
                .carrierDistanceSq(SummonAiNumbers.FOLLOW_CARRIER_DISTANCE_SQ + 0.01D)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldFollowCarrier(stateBuilder()
                .carrierDistanceSq(SummonAiNumbers.FOLLOW_CARRIER_DISTANCE_SQ)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldFollowCarrier(stateBuilder()
                .carrierPresent(false)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldFollowCarrier(stateBuilder()
                .carrierAlive(false)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldFollowCarrier(stateBuilder()
                .hasAttackTarget(true)
                .build()));
    }

    @Test
    void randomStrollShouldNotRaceCarrierFollowOrCombatMovement() {
        assertTrue(SummonBrainDecisionResolver.shouldRandomStroll(stateBuilder()
                .carrierDistanceSq(0.0D)
                .randomStrollRollHit(true)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldRandomStroll(stateBuilder()
                .carrierPresent(false)
                .randomStrollRollHit(true)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldRandomStroll(stateBuilder()
                .carrierAlive(false)
                .randomStrollRollHit(true)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldRandomStroll(stateBuilder()
                .hasAttackTarget(true)
                .randomStrollRollHit(true)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldRandomStroll(stateBuilder()
                .carrierDistanceSq(SummonAiNumbers.FOLLOW_CARRIER_DISTANCE_SQ + 0.01D)
                .randomStrollRollHit(true)
                .build()));
        assertFalse(SummonBrainDecisionResolver.shouldRandomStroll(stateBuilder()
                .randomStrollRollHit(false)
                .build()));
    }

    private static StateBuilder stateBuilder() {
        return new StateBuilder();
    }

    private static final class StateBuilder {
        private boolean carrierPresent = true;
        private boolean carrierAlive = true;
        private double carrierDistanceSq;
        private boolean hasAttackTarget;
        private boolean hasAliveAttackTarget;
        private boolean randomStrollRollHit;
        private double targetDistanceSq;
        private double attackRangeSq = SummonAiNumbers.DEFAULT_ATTACK_RANGE_SQ;
        private int attackDelay;

        private StateBuilder carrierPresent(boolean carrierPresent) {
            this.carrierPresent = carrierPresent;
            return this;
        }

        private StateBuilder carrierAlive(boolean carrierAlive) {
            this.carrierAlive = carrierAlive;
            return this;
        }

        private StateBuilder carrierDistanceSq(double carrierDistanceSq) {
            this.carrierDistanceSq = carrierDistanceSq;
            return this;
        }

        private StateBuilder hasAttackTarget(boolean hasAttackTarget) {
            this.hasAttackTarget = hasAttackTarget;
            return this;
        }

        private StateBuilder hasAliveAttackTarget(boolean hasAliveAttackTarget) {
            this.hasAliveAttackTarget = hasAliveAttackTarget;
            return this;
        }

        private StateBuilder randomStrollRollHit(boolean randomStrollRollHit) {
            this.randomStrollRollHit = randomStrollRollHit;
            return this;
        }

        private StateBuilder targetDistanceSq(double targetDistanceSq) {
            this.targetDistanceSq = targetDistanceSq;
            return this;
        }

        private StateBuilder attackRangeSq(double attackRangeSq) {
            this.attackRangeSq = attackRangeSq;
            return this;
        }

        private StateBuilder attackDelay(int attackDelay) {
            this.attackDelay = attackDelay;
            return this;
        }

        private SummonBrainDecisionResolver.State build() {
            return new SummonBrainDecisionResolver.State(
                    this.carrierPresent,
                    this.carrierAlive,
                    this.carrierDistanceSq,
                    this.hasAttackTarget,
                    this.hasAliveAttackTarget,
                    this.randomStrollRollHit,
                    this.targetDistanceSq,
                    this.attackRangeSq,
                    this.attackDelay
            );
        }
    }
}

File: src/test/kotlin/org/trp/shincolle/entity/base/SummonBrainDecisionResolverTest.kt
package org.trp.shincolle.entity.base

import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class SummonBrainDecisionResolverTest {

    @Test
    fun `attack should require alive target`() {
        assertThat(
            SummonBrainDecisionResolver.shouldAttack(
                stateBuilder().hasAttackTarget(true).hasAliveAttackTarget(true).build()
            )
        ).isTrue()
        assertThat(
            SummonBrainDecisionResolver.shouldAttack(
                stateBuilder().hasAttackTarget(true).hasAliveAttackTarget(false).build()
            )
        ).isFalse()
    }

    @Test
    fun `chase should only continue while target is alive and out of range`() {
        assertThat(
            SummonBrainDecisionResolver.shouldChaseAttackTarget(
                stateBuilder()
                    .hasAttackTarget(true)
                    .hasAliveAttackTarget(true)
                    .targetDistanceSq(16.01)
                    .attackRangeSq(16.0)
                    .build()
            )
        ).isTrue()
        assertThat(
            SummonBrainDecisionResolver.shouldChaseAttackTarget(
                stateBuilder()
                    .hasAttackTarget(true)
                    .hasAliveAttackTarget(true)
                    .targetDistanceSq(16.0)
                    .attackRangeSq(16.0)
                    .build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldChaseAttackTarget(
                stateBuilder()
                    .hasAttackTarget(true)
                    .hasAliveAttackTarget(false)
                    .targetDistanceSq(20.0)
                    .attackRangeSq(16.0)
                    .build()
            )
        ).isFalse()
    }

    @Test
    fun `perform attack should require alive target cooldown and in range`() {
        assertThat(
            SummonBrainDecisionResolver.shouldPerformAttack(
                stateBuilder()
                    .hasAttackTarget(true)
                    .hasAliveAttackTarget(true)
                    .targetDistanceSq(16.0)
                    .attackRangeSq(16.0)
                    .attackDelay(0)
                    .build()
            )
        ).isTrue()
        assertThat(
            SummonBrainDecisionResolver.shouldPerformAttack(
                stateBuilder()
                    .hasAttackTarget(true)
                    .hasAliveAttackTarget(true)
                    .targetDistanceSq(16.01)
                    .attackRangeSq(16.0)
                    .attackDelay(0)
                    .build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldPerformAttack(
                stateBuilder()
                    .hasAttackTarget(true)
                    .hasAliveAttackTarget(true)
                    .targetDistanceSq(16.0)
                    .attackRangeSq(16.0)
                    .attackDelay(1)
                    .build()
            )
        ).isFalse()
    }

    @Test
    fun `follow carrier should require alive carrier distance and no attack target`() {
        assertThat(
            SummonBrainDecisionResolver.shouldFollowCarrier(
                stateBuilder().carrierDistanceSq(SummonAiNumbers.FOLLOW_CARRIER_DISTANCE_SQ + 0.01).build()
            )
        ).isTrue()
        assertThat(
            SummonBrainDecisionResolver.shouldFollowCarrier(
                stateBuilder().carrierDistanceSq(SummonAiNumbers.FOLLOW_CARRIER_DISTANCE_SQ).build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldFollowCarrier(
                stateBuilder().carrierPresent(false).build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldFollowCarrier(
                stateBuilder().carrierAlive(false).build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldFollowCarrier(
                stateBuilder().hasAttackTarget(true).build()
            )
        ).isFalse()
    }

    @Test
    fun `random stroll should not race carrier follow or combat movement`() {
        assertThat(
            SummonBrainDecisionResolver.shouldRandomStroll(
                stateBuilder().carrierDistanceSq(0.0).randomStrollRollHit(true).build()
            )
        ).isTrue()
        assertThat(
            SummonBrainDecisionResolver.shouldRandomStroll(
                stateBuilder().carrierPresent(false).randomStrollRollHit(true).build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldRandomStroll(
                stateBuilder().carrierAlive(false).randomStrollRollHit(true).build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldRandomStroll(
                stateBuilder().hasAttackTarget(true).randomStrollRollHit(true).build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldRandomStroll(
                stateBuilder()
                    .carrierDistanceSq(SummonAiNumbers.FOLLOW_CARRIER_DISTANCE_SQ + 0.01)
                    .randomStrollRollHit(true)
                    .build()
            )
        ).isFalse()
        assertThat(
            SummonBrainDecisionResolver.shouldRandomStroll(
                stateBuilder().randomStrollRollHit(false).build()
            )
        ).isFalse()
    }

    private fun stateBuilder(): StateBuilder = StateBuilder()

    private class StateBuilder {
        private var carrierPresent = true
        private var carrierAlive = true
        private var carrierDistanceSq = 0.0
        private var hasAttackTarget = false
        private var hasAliveAttackTarget = false
        private var randomStrollRollHit = false
        private var targetDistanceSq = 0.0
        private var attackRangeSq = SummonAiNumbers.DEFAULT_ATTACK_RANGE_SQ.toDouble()
        private var attackDelay = 0

        fun carrierPresent(carrierPresent: Boolean) = apply { this.carrierPresent = carrierPresent }
        fun carrierAlive(carrierAlive: Boolean) = apply { this.carrierAlive = carrierAlive }
        fun carrierDistanceSq(carrierDistanceSq: Double) = apply { this.carrierDistanceSq = carrierDistanceSq }
        fun hasAttackTarget(hasAttackTarget: Boolean) = apply { this.hasAttackTarget = hasAttackTarget }
        fun hasAliveAttackTarget(hasAliveAttackTarget: Boolean) = apply { this.hasAliveAttackTarget = hasAliveAttackTarget }
        fun randomStrollRollHit(randomStrollRollHit: Boolean) = apply { this.randomStrollRollHit = randomStrollRollHit }
        fun targetDistanceSq(targetDistanceSq: Double) = apply { this.targetDistanceSq = targetDistanceSq }
        fun attackRangeSq(attackRangeSq: Double) = apply { this.attackRangeSq = attackRangeSq }
        fun attackDelay(attackDelay: Int) = apply { this.attackDelay = attackDelay }

        fun build(): SummonBrainDecisionResolver.State {
            return SummonBrainDecisionResolver.State(
                carrierPresent,
                carrierAlive,
                carrierDistanceSq,
                hasAttackTarget,
                hasAliveAttackTarget,
                randomStrollRollHit,
                targetDistanceSq,
                attackRangeSq,
                attackDelay
            )
        }
    }
}
